File: problems/basic/622/solution.py
# Accepted: 62ms (79.34%), 17.97MB (20.49%)

class MyCircularQueue:
    def __init__(self, k: int):
        self.internalSize = k+1 # reserver extra one block to detect whether the queue is full
        self.array = [0] * self.internalSize
        self.toPush = 0
        self.toPop = 0

    def _getPrevIndex(self, index):
        return (index + self.internalSize-1) % self.internalSize

    def _getNextIndex(self, index):
        return (index+1) % self.internalSize

    def enQueue(self, value: int) -> bool:
        if self.isFull():
            return False

        self.array[self.toPush] = value
        self.toPush = self._getNextIndex(self.toPush)
        return True

    def deQueue(self) -> bool:
        if self.isEmpty():
            return False

        self.toPop = self._getNextIndex(self.toPop)
        return True

    def Front(self) -> int:
        if self.isEmpty():
            return -1

        return self.array[self.toPop]

    def Rear(self) -> int:
        if self.isEmpty():
            return -1

        rear = self._getPrevIndex(self.toPush)
        return self.array[rear]

    def isEmpty(self) -> bool:
        return self.toPush == self.toPop

    def isFull(self) -> bool:
        return self._getNextIndex(self.toPush) == self.toPop

# Your MyCircularQueue object will be instantiated and called as such:
# obj = MyCircularQueue(k)
# param_1 = obj.enQueue(value)
# param_2 = obj.deQueue()
# param_3 = obj.Front()
# param_4 = obj.Rear()
# param_5 = obj.isEmpty()
# param_6 = obj.isFull()

File: problems/basic/622/solution.cpp
// Accepted: 23ms (38.64%), 17.32MB (8.22%)

class MyCircularQueue {
    int internalSize;
    vector<int> array;
    int toPush;
    int toPop;

    int getNextIndex(int i) {
        return (i+1) % internalSize;
    }
    int getPrevIndex(int i) {
        return (i+internalSize-1) % internalSize;
    }

public:
    MyCircularQueue(int k) {
        internalSize = k+1;
        for (int i = 0; i < internalSize; ++i) {
            array.push_back(0);
        }
        toPush = toPop = 0;
    }

    bool enQueue(int value) {
        if (isFull()) {
            return false;
        }

        array[toPush] = value;
        toPush = getNextIndex(toPush);
        return true;
    }

    bool deQueue() {
        if (isEmpty()) {
            return false;
        }

        toPop = getNextIndex(toPop);
        return true;
    }

    int Front() {
        if (isEmpty()) {
            return -1;
        }

        return array[toPop];
    }

    int Rear() {
        if (isEmpty()) {
            return -1;
        }

        int rear = getPrevIndex(toPush);
        return array[rear];
    }

    bool isEmpty() {
        return toPush == toPop;
    }

    bool isFull() {
        return getNextIndex(toPush) == toPop;
    }
};

/**
 * Your MyCircularQueue object will be instantiated and called as such:
 * MyCircularQueue* obj = new MyCircularQueue(k);
 * bool param_1 = obj->enQueue(value);
 * bool param_2 = obj->deQueue();
 * int param_3 = obj->Front();
 * int param_4 = obj->Rear();
 * bool param_5 = obj->isEmpty();
 * bool param_6 = obj->isFull();
 */
